ambari-server/src/main/java/org/apache/ambari/server/api/services/mpackadvisor/MpackAdvisorBlueprintProcessor.java

```diff
@@ -0,0 +1,334 @@
+    for (Map.Entry<String, HostGroupInfo> hgEnrty: clusterTopology.getHostGroupInfo().entrySet()) {
+      hgBindngs.put(hgEnrty.getKey(), Sets.newCopyOnWriteArraySet(hgEnrty.getValue().getHostNames()));
+    }
+    return hgBindngs;
+  }
+
+  private Map<String, Set<Component>> gatherHostGroupComponents(ClusterTopology clusterTopology) {
+    Map<String, Set<Component>> hgComponentsMap = Maps.newHashMap();
+    for (Map.Entry<String, HostGroup> hgEnrty: clusterTopology.getBlueprint().getHostGroups().entrySet()) {
+      hgComponentsMap.put(hgEnrty.getKey(), Sets.newCopyOnWriteArraySet(hgEnrty.getValue().getComponents()));
+    }
+    return hgComponentsMap;
+  }
+
+  private Map<String, Map<String, Map<String, String>>> calculateConfigs(Configuration configuration) {
+    Map<String, Map<String, Map<String, String>>> result = Maps.newHashMap();
+    Map<String, Map<String, String>> fullProperties = configuration.getFullProperties();
+    for (Map.Entry<String, Map<String, String>> siteEntry : fullProperties.entrySet()) {
+      Map<String, Map<String, String>> propsMap = Maps.newHashMap();
+      propsMap.put("properties", siteEntry.getValue());
+      result.put(siteEntry.getKey(), propsMap);
+    }
+    return result;
+  }
+
+  private Map<String, Map<String, Set<String>>> gatherMackComponentsHostsMap(Map<String, Set<Component>> hostGroupComponents,
+                                                                Map<String, Set<String>> hostGroupHosts,
+                                                                Set<MpackInstance> mpacks) {
+
+    // Calculate helper maps
+    Map<String, Set<String>> mpackComponents = mpacks.stream()
+      .collect(toMap(
+        MpackInstance::getMpackType, // TODO: this is supposed to be MpackInstance::getMpackName, fix mpack advisor
+        mpack -> getStackComponents(mpack.getStackId())
+      ));
+
+    Map<String, Set<String>> componentNameToMpacks = mpackComponents.entrySet().stream()
+      .flatMap( mpc -> mpc.getValue().stream().map(component -> Pair.of(component, ImmutableSet.of(mpc.getKey()))) )
+      .collect(toMap(
+        Pair::getLeft,
+        Pair::getRight,
+        (set1, set2) -> ImmutableSet.copyOf(Sets.union(set1, set2))
+      ));
+
+    Map<String, Set<Component>> hostComponents = hostGroupComponents.entrySet().stream()
+      .flatMap( hgc -> hostGroupHosts.get(hgc.getKey()).stream().map(host -> Pair.of(host, hgc.getValue())))
+      .collect(toMap(Pair::getLeft, Pair::getRight));
+
+    // Calculate map to return: mpack -> component -> host
+    Map<String, Map<String, Set<String>>> mpackComponentHostsMap = new HashMap<>();
+
+    hostComponents.entrySet().forEach( entry -> {
+      String hostName = entry.getKey();
+      Set<Component> components = entry.getValue();
+      components.forEach( component -> {
+        Set<String> mpacksForComponent = getMpacksForComponent(component, componentNameToMpacks);
+        mpacksForComponent.forEach(mpack -> {
```

Comment:
The way this class associates components to mpacks (and services in `copyAndEnrichMpackInstances`) is a bit redundant, since it's already performed by [`StackComponentResolver`](https://github.com/apache/ambari/blob/branch-feature-AMBARI-14714/ambari-server/src/main/java/org/apache/ambari/server/topology/StackComponentResolver.java) and the result is stored in [`ResolvedComponent`](https://github.com/apache/ambari/blob/branch-feature-AMBARI-14714/ambari-server/src/main/java/org/apache/ambari/server/topology/ResolvedComponent.java) instances.  If some information is missing, I think it should be added there to avoid duplicated, and potentially different, implementations.

Reply:
Good catch. I've rewritten the logic to take advantage of ResolvedComponent's.